c++ 17 for Windows (msvc)?

Hi,

I'm trying to build `rmw_iceoryx` on Win11 with Visual studio 2022 and I had a compiling error.

```
C:/ros/humble/install/include/iceoryx/v2.0.3\iceoryx_hoofs/cxx/type_traits.hpp(63,28): error C2039: 'invoke_result': is not a member of 'std' 
(compiling source file C:\ros\humble\src\ros2\rmw_iceoryx\rmw_iceoryx_cpp\src\rmw_init.cpp) 
[C:\ros\humble\src\ros2\rmw_iceoryx\rmw_iceoryx_cpp\build\rmw_iceoryx_cpp\rmw_iceoryx_cpp.vcxproj]
```

the code is actually from `iceoryx` itself but it's included in `rmw_init.cpp`. The source code `type_traits.hpp(63,28)` is:
```c++
// windows defines __cplusplus as 199711L
#if __cplusplus < 201703L && !defined(_WIN32)
template <typename C, typename... Cargs>
using invoke_result = std::result_of<C(Cargs...)>;
#elif __cplusplus >= 201703L || defined(_WIN32)
template <typename C, typename... Cargs>
using invoke_result = std::invoke_result<C, Cargs...>;
#endif
```
In my case, my  `__cplusplus` is greater than `201703L` and so ` std::invoke_result` is used.

But ` std::invoke_result` is a feature in **c++ 17**.

The current project is c++ 14 as in cmake https://github.com/ros2/rmw_iceoryx/blob/humble/rmw_iceoryx_cpp/CMakeLists.txt#LL9C1-L12C8:
```
# Default to C++14
if(NOT CMAKE_CXX_STANDARD)
  set(CMAKE_CXX_STANDARD 14)
endif()
```

the code above is flexible and I can set `CMAKE_CXX_STANDARD `.

Then I checked the source code of `iceoryx` itself, 

the older version (v2.0.3 as in ros2 humble) of code is :
```
if(LINUX)
    set(ICEORYX_CXX_STANDARD 14)
elseif(QNX)
    set(ICEORYX_CXX_STANDARD 14)
elseif(WIN32)
    set(ICEORYX_CXX_STANDARD 17)
elseif(APPLE)
    set(ICEORYX_CXX_STANDARD 17)
elseif(UNIX)
   set(ICEORYX_CXX_STANDARD 17)
else()
   set(ICEORYX_CXX_STANDARD 17)
endif()
```
and the latest code is moved to https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_platform/win/IceoryxPlatformSettings.cmake#L18 and is:
```cmake
set_global(VAR ICEORYX_PLATFORM_STRING      VALUE "Windows")
set_global(VAR ICEORYX_CXX_STANDARD         VALUE 17)
```

So, my question is: does it make sense to set c++ 17 in windows build just like in `iceoryx`?

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -5,11 +5,11 @@ on:
 jobs:
   lint:
     name: ament_${{ matrix.linter }}
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-22.04
     strategy:
       fail-fast: false
       matrix:
-          linter: [cppcheck, cpplint, uncrustify, xmllint, copyright]
+          linter: [cpplint, uncrustify, xmllint, copyright]
     steps:
     - uses: actions/checkout@v2
     - uses: ros-tooling/setup-ros@master

--- a/rmw_iceoryx_cpp/CMakeLists.txt
+++ b/rmw_iceoryx_cpp/CMakeLists.txt
@@ -6,9 +6,20 @@ if(NOT CMAKE_C_STANDARD)
   set(CMAKE_C_STANDARD 99)
 endif()
 
-# Default to C++14
-if(NOT CMAKE_CXX_STANDARD)
+# Set C++ standard in accordance with iceoryx
+# https://github.com/eclipse-iceoryx/iceoryx/blob/v2.0.3/iceoryx_hoofs/cmake/IceoryxPlatform.cmake#L26-L38
+if(LINUX)
   set(CMAKE_CXX_STANDARD 14)
+elseif(QNX)
+  set(CMAKE_CXX_STANDARD 14)
+elseif(WIN32)
+  set(CMAKE_CXX_STANDARD 17)
+elseif(APPLE)
+  set(CMAKE_CXX_STANDARD 17)
+elseif(UNIX)
+  set(CMAKE_CXX_STANDARD 17)
+else()
+  set(CMAKE_CXX_STANDARD 17)
 endif()
 
 if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")

